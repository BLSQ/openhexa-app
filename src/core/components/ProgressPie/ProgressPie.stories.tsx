import { Story } from "@ladle/react";

import ProgressPie, { ProgressPieProps } from "./ProgressPie";

interface ProgressPieStoryProps extends ProgressPieProps {}

export const ProgressPieStory: Story<ProgressPieStoryProps> = () => {
  return (
    <div>
      <ProgressPie progress={10} />
      <ProgressPie progress={30} size={50} />
    </div>
  );
};

ProgressPieStory.storyName = "ProgressPie";

const defaults = {
  title: "UI / ProgressPie",
};
export default defaults;
