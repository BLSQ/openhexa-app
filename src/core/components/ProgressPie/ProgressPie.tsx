import clsx from "clsx";
import { useMemo } from "react";

export type ProgressPieProps = {
  progress: number;
  background?: string;
  foreground?: string;
  textClassName?: string;
  className?: string;
};

const ProgressPie = (props: ProgressPieProps) => {
  const {
    progress,
    background = "stroke-gray-100",
    foreground = "stroke-gray-400",
    textClassName = "text-gray-400",
    className,
  } = props;

  const value = useMemo(() => Math.min(100, Math.max(0, progress)), [progress]);
  return (
    <div
      className={clsx("relative flex items-center justify-center", className)}
    >
      <svg
        id="svg"
        width="auto"
        height="auto"
        viewBox="-1 -1 34 34"
        className="absolute inset-0 -rotate-90"
      >
        <circle
          cx="16"
          cy="16"
          r="15.9155"
          fill="none"
          strokeWidth="2"
          className={background}
        />

        <circle
          cx="16"
          cy="16"
          r="15.9155"
          strokeDasharray={"100 100"}
          strokeDashoffset={100 - value}
          strokeLinecap="round"
          strokeWidth="2"
          fill="none"
          className={clsx(
            "transition-all duration-700 ease-in-out ",
            foreground
          )}
        />
      </svg>
      {value !== 0 && (
        <div
          className={clsx("z-10 font-mono text-xs font-normal", textClassName)}
        >
          {value}%
        </div>
      )}
    </div>
  );
};

export default ProgressPie;
