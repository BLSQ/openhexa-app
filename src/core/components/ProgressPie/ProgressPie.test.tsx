import { render } from "@testing-library/react";
import ProgressPie from "./ProgressPie";

describe("ProgressPie", () => {
  it("renders a progress pie at different progress states", async () => {
    const { container, rerender } = render(<ProgressPie progress={0} />);
    expect(container).toMatchSnapshot();

    rerender(<ProgressPie progress={50} />);
    expect(container).toMatchSnapshot();

    rerender(<ProgressPie progress={100} />);
    expect(container).toMatchSnapshot();
  });

  it("accepts a different set of colours", () => {
    const { container } = render(
      <ProgressPie
        progress={30}
        background="text-teal-500"
        foreground="text-teal-700"
      />,
    );

    expect(container.getElementsByClassName("text-teal-500").length).toEqual(1);
    expect(container.getElementsByClassName("text-teal-700").length).toEqual(1);
  });
});
