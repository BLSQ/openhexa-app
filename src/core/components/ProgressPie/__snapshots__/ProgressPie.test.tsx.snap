// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressPie renders a progress pie at different progress states 1`] = `
<div>
  <div
    class="relative flex items-center justify-center"
  >
    <svg
      class="absolute inset-0 -rotate-90"
      id="svg"
      viewBox="-1 -1 34 34"
    >
      <circle
        class="stroke-gray-100"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-width="2"
      />
      <circle
        class="transition-all duration-700 ease-in-out  stroke-gray-400"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-dasharray="100 100"
        stroke-dashoffset="100"
        stroke-linecap="round"
        stroke-width="2"
      />
    </svg>
    <div
      class="z-10 font-mono text-xs font-normal text-gray-400"
    >
      0
      %
    </div>
  </div>
</div>
`;

exports[`ProgressPie renders a progress pie at different progress states 2`] = `
<div>
  <div
    class="relative flex items-center justify-center"
  >
    <svg
      class="absolute inset-0 -rotate-90"
      id="svg"
      viewBox="-1 -1 34 34"
    >
      <circle
        class="stroke-gray-100"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-width="2"
      />
      <circle
        class="transition-all duration-700 ease-in-out  stroke-gray-400"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-dasharray="100 100"
        stroke-dashoffset="50"
        stroke-linecap="round"
        stroke-width="2"
      />
    </svg>
    <div
      class="z-10 font-mono text-xs font-normal text-gray-400"
    >
      50
      %
    </div>
  </div>
</div>
`;

exports[`ProgressPie renders a progress pie at different progress states 3`] = `
<div>
  <div
    class="relative flex items-center justify-center"
  >
    <svg
      class="absolute inset-0 -rotate-90"
      id="svg"
      viewBox="-1 -1 34 34"
    >
      <circle
        class="stroke-gray-100"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-width="2"
      />
      <circle
        class="transition-all duration-700 ease-in-out  stroke-gray-400"
        cx="16"
        cy="16"
        fill="none"
        r="15.9155"
        stroke-dasharray="100 100"
        stroke-dashoffset="0"
        stroke-linecap="round"
        stroke-width="2"
      />
    </svg>
    <div
      class="z-10 font-mono text-xs font-normal text-gray-400"
    >
      100
      %
    </div>
  </div>
</div>
`;
